### PR TITLE
GatewayClass with missing parametersRef or missing fields is validated

### PR DESCRIPTION
**Pull request by sunjayBhatia**
Adds unit tests to cover this case

Removes some nil checks in internal code

Updates #337

### CODE REVIEW

**Comment by sunjayBhatia** [comment]
@danehans i see you opened https://github.com/projectcontour/contour-operator/issues/337

The docs say `This is optional if the controller does not require any additional configuration.`

Is there any case where one can actually create a Contour without a Contour CRD instance? If so, seems like we need some processing step to fill in values and some more testing since there are pointers all over the place

**Comment by danehans** [comment]
@sunjayBhatia my thinking was that if paramRef was unspecified, Contour is provisioned with default settings, i.e. Envoy LB service, etc.. Thoughts @youngnick @stevesloka @skriss?

File: pkg/validation/validation.go [hunk 1]
```diff
@@ -109,18 +109,20 @@ func nodePorts(contour *operatorv1alpha1.Contour) error {
...
 	return parameterRef(gc)
 }
 
-// parameterRef returns true if parametersRef of gc is valid.
+// parameterRef returns nil if parametersRef of gc is valid,
+// otherwise an error.
 func parameterRef(gc *gatewayv1alpha1.GatewayClass) error {
 	if gc.Spec.ParametersRef == nil {
-		return nil
+		return fmt.Errorf("invalid gateway class %s, missing parametersRef", gc.Name)
```
Comment (danehans) [inline pkg/validation/validation.go:122]: gatewayclass should be one word, similar to daemonset, ingressclass, etc..

File: pkg/validation/validation.go [hunk 2]
```diff
@@ -109,18 +109,20 @@ func nodePorts(contour *operatorv1alpha1.Contour) error {
...
 	}
 	if gc.Spec.ParametersRef.Scope == nil || *gc.Spec.ParametersRef.Scope != gatewayClassNamespacedParamRef {
-		return fmt.Errorf("invalid parametersRef for gateway class %s, only namespaced-scoped referecnes are supported", gc.Name)
+		return fmt.Errorf("invalid parametersRef for gateway class %s, only namespaced-scoped references are supported", gc.Name)
```
Comment (danehans) [inline pkg/validation/validation.go:125]: same as https://github.com/projectcontour/contour-operator/pull/339/files#r625960661

File: pkg/validation/validation.go [hunk 3]
```diff
@@ -130,6 +132,9 @@ func parameterRef(gc *gatewayv1alpha1.GatewayClass) error {
 	if kind != operatorv1alpha1.GatewayClassParamsRefKind {
 		return fmt.Errorf("invalid kind %q", kind)
 	}
+	if gc.Spec.ParametersRef.Namespace == nil {
+		return fmt.Errorf("invalid parametersRef for gateway class %s, missing namespace", gc.Name)
```
Comment (danehans) [inline pkg/validation/validation.go:136]: same as https://github.com/projectcontour/contour-operator/pull/339/files#r625960661

**Comment by danehans** [comment]
@sunjayBhatia I'm fine proceeding with this PR as-is, but I would like an issue created and xref'd for handling a gc with no paramRef specified. In this scenario, the gateway would be treated the same as a default Contour. For example:
```
kind: GatewayClass
apiVersion: networking.x-k8s.io/v1alpha1
metadata:
  name: sample-gatewayclass
spec:
  controller: projectcontour.io/contour-operator
---
kind: Gateway
apiVersion: networking.x-k8s.io/v1alpha1
metadata:
  name: contour
  namespace: projectcontour
spec:
  gatewayClassName: sample-gatewayclass
  ...
```

The above is the same as:
```
apiVersion: operator.projectcontour.io/v1alpha1
kind: Contour
metadata:
  name: contour-sample
spec: {}
```

**Comment by sunjayBhatia** [comment]
Sounds like this issue you created? https://github.com/projectcontour/contour-operator/issues/337

This PR was mainly to move the verbose nil checks into the validation package so we don't have to repeat them all over and to fix any crashes that could arise from fields not being checked after a resource is admitted.

I think it makes sense to support the use case you describe. This change restricts it for now and should solve panics etc. that are caused by things in a gatewayclass being nil etc. but once we refactor we can enable a nil parametersRef field/default Contour etc.

I think the issue of having to do nil checks etc. all over the code vs. rejecting resources with missing fields is an interesting one. I think because we're maybe not translating k8s resources into an internal abstraction type we have to repeat a lot of code to do nil checks etc. Could be a good way to refactor the pipeline: validate resource, translate it to internal type if valid (filling in any missing fields with defaults), use to generate further config.

**Comment by danehans** [comment]
CI failing due to #341.

**Comment by danehans** [comment]
> This PR was mainly to move the verbose nil checks into the validation package so we don't have to repeat them all over and to fix any crashes that could arise from fields not being checked after a resource is admitted.

I think it makes sense to support the use case you describe. This change restricts it for now and should solve panics etc. that are caused by things in a gatewayclass being nil etc. but once we refactor we can enable a nil parametersRef field/default Contour etc.

+1

**Review by danehans — APPROVED** [review]

**Pull request merged**